zerver/lib/send_email.py

```diff
@@ -231,6 +246,23 @@ def send_email_from_dict(email_dict: Mapping[str, Any]) -> None:
     send_email(**dict(email_dict))
 
 
+@backoff.on_exception(backoff.expo, OSError, max_tries=MAX_CONNECTION_TRIES)
+def initialize_connection(connection: Optional[BaseEmailBackend]) -> BaseEmailBackend:
+    if not connection:
+        connection = get_connection()
+    connection.open()
+    # Ensures that we don't return a connection that has been closed by the mail server
+    if isinstance(connection, EmailBackend):
+        try:
+            status = connection.connection.noop()[0]
+        except Exception:
+            status = -1
+        if status != 250:
+            # Try again; calling ourselves without passing down the 'connection' will make a new low-level SMTP connection
+            return initialize_connection()
+    return connection
```

Comment:
Looking at my suggested code with fresh eyes, I think it has the failure mode that if a `noop` is _unsuccessful after a new connection is opened_ then we'll recurse infinitely here, busily making and trying to `noop` connections.

The other problem is that, from briefly poking at it, `@backoff.on_exception` re-calls its decorated function with the same parameters.  And a `connection` which has already failed its low-level noop, but still thinks it's open, is going to not re-open the TCP connection, and fail repeatedly.  Recursing is also going to open us up to the problem where the inner call retries 3 times, then re-raises -- whereupon the _outer_ call will retry 3 times.

Ditching the recursive calls, and being explicit about open/close, is I think enough here to fix this:
```suggestion
    if connection.open():
        # If it's a new connection, no need to no-op to check connectivity
        return connection
    # No-op to ensures that we don't return a connection that has been closed by the mail server
    if isinstance(connection, EmailBackend):
        try:
            status = connection.connection.noop()[0]
        except Exception:
            status = -1
        if status != 250:
            # Close and connect again.
            connection.close()
            connection.open()
    return connection
```

Reply:
Indeed these changes seem to be required for this function to work properly. Modified.